12-http/example3/main.go

```diff
@@ -18,20 +21,46 @@ type user struct {
 	Phone string
 }
 
-// users is a slice of users.
-var users []user
+var (
+	// mu protects users.
+	mu sync.RWMutex
+	// initialize users, so that the output JSON won't be null
+	// prior to adding users.
+	users = make([]user, 0)
+)
+
+// index page template
+var IdxTpl = template.Must(template.ParseFiles("template/index.html"))
+
+// file server for any future assets and other static files
+var fs = http.FileServer(http.Dir("public"))
 
 func main() {
+	// This will handle all paths without specific routes
+	http.HandleFunc("/", baseHandler)
+
 	// Create a ServeMux and add some routes.
-	mux := http.NewServeMux()
-	mux.HandleFunc("/users", usersHandler)
-	mux.HandleFunc("/search", searchUsers)
+	api := http.NewServeMux()
+	api.HandleFunc("/users", usersHandler)
+	api.HandleFunc("/search", searchUsers)
 
-	// This will act as a catch all for the rest of the routes
-	mux.Handle("/", http.FileServer(http.Dir("public")))
+	http.Handle("/api/v1/", http.StripPrefix("/api/v1", api))
 
 	// Start the service.
-	http.ListenAndServe(":4000", mux)
+	bind := ":4000"
+	log.Println("Serving HTTP on", bind)
+	log.Fatalln(http.ListenAndServe(bind, nil))
+}
+
+func baseHandler(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Path != "/" {
+		fs.ServeHTTP(w, r)
```

Comment:
Yeah. Stylistically do the shorter-code thing first (in this case, falling back to static file serving if the path isn't exactly "/"). Since we're using an explicit Mux on /api/v1 in func main, then anything with an /api/v1 prefix won't get handled by baseHandler.